src/ocean/util/container/btree/BTreeRange.d

```diff
@@ -0,0 +1,355 @@
+/********************************************************************************
+
+    Range traversal support for BTree.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.util.container.btree.BTreeRange;
+
+import ocean.transition;
+import ocean.core.array.Mutation;
+import ocean.util.container.btree.BTree;
+
+/*******************************************************************************
+
+    Exception thrown from the BTreeRange if the iteration over the changed
+    tree is performed.
+
+*******************************************************************************/
+
+class RangeInvalidatedException: Exception
+{
+    import ocean.core.Exception: DefaultExceptionCtor;
+    mixin DefaultExceptionCtor;
+}
+
+// shared between different ranges
+private static RangeInvalidatedException range_exception;
+
+static this ()
+{
+    .range_exception = new RangeInvalidatedException(
+            "Container has been changed during iteration, please restart.");
+}
+
+/*******************************************************************************
+
+    Returns a range that iterates the tree via the inorder.
+
+    If tree has changed during the iteration, range will throw an
+    RangeInvalidatedException. If the user knows that the tree will not change
+    during the range traversal, it is safe to call range's methods - it will throw
+    an exception if this promise is broken. In case user knows that the tree
+    can be changed any time, it is possible to call BTreeRange.isValid before
+    performing any action on the range to check if the accessing is safe. This
```

Comment:
I'll clarify more. It means that either fetching the current element, or moving to the next one is guaranteed to succeed in a single-thread environment.